clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java

```diff
@@ -3437,21 +3443,318 @@ public void testPollWithRedundantCreateFetchRequests() {
+        assertEquals(1, sendFetches());
+    }
+
+    @Test
+    public void testFetchRequestWithBufferedPartitionMissingPosition() {
+        buildFetcher();
+
+        // The test requires that there are multiple nodes as the fetch request logic is based in part off of the
+        // partition-to-node relationship.
+        int numNodes = 2;
+        Set<TopicPartition> partitions = Set.of(tp0, tp1, tp2, tp3);
+        assignFromUser(partitions, numNodes);
+
+        // Get all the nodes serving as the leader for these partitions, and then split them into the separate
+        // nodes and set of partitions to make things easier to keep track of later.
+        List<Node> nodes = nodesForPartitionLeaders(partitions);
+        Node node0 = nodes.get(0);
+        Node node1 = nodes.get(1);
+        List<TopicPartition> node0Partitions = partitionsForNode(node0, partitions);
+        List<TopicPartition> node1Partitions = partitionsForNode(node1, partitions);
+
+        // Seek each partition so that it becomes eligible to fetch.
+        partitions.forEach(tp -> subscriptions.seek(tp, 0));
+
+        // sendFetches() call #1 should issue requests to node 0 or node 1 since neither has buffered data.
+        assertEquals(2, sendFetches());
+        prepareFetchResponses(node0, node0Partitions, 0);
+        prepareFetchResponses(node1, node1Partitions, 0);
+        networkClientDelegate.poll(time.timer(0));
+
+        // Grab both partitions for node 0. The first partition will be collected so that it doesn't have anything
+        // in the fetch buffer. The second node will be left in the buffer, but will clear out its position's leader
+        // node.
+        TopicPartition node0Partition1 = node0Partitions.remove(0);
+        TopicPartition node0Partition2 = node0Partitions.remove(0);
+
+        assertEquals(4, fetcher.fetchBuffer.bufferedPartitions().size());
+        collectSelectedPartition(node0Partition1, partitions);
+        assertEquals(3, fetcher.fetchBuffer.bufferedPartitions().size());
+
+        // Overwrite the position with an empty leader to trigger the test case.
+        subscriptions.position(node0Partition2, null);
+
+        // Both the collected partition and the position without a partition leader should have a retrievable position.
+        // Confirm that position() doesn't throw an exception and that the leader for the second partition is missing.
+        assertNotEquals(Optional.empty(), subscriptions.position(node0Partition1).currentLeader.leader);
+        assertNull(subscriptions.position(node0Partition2));
+
+        // sendFetches() call #2 will now fail to send any requests as we have an invalid position in the assignment.
+        // The Consumer.poll() API will throw an IllegalStateException to the user.
+        Future<Void> future = fetcher.createFetchRequests();
+        assertEquals(0, sendFetches());
+        assertFutureThrows(future, IllegalStateException.class);
```

Comment:
Is this the right behavior? The original code only calls `subscriptions.position` for partitions that are fetchable, which implies that the position is available. So, it will never throw IllegalStateException in this case.

Reply:
The original code from `AbstractFetch.prepareFetchRequests()` checked for a `null` `FetchPosition` even though `SubscriptionState.fetchablePartitions()` only returns fetchable partitions. I assumed the existing code was there to handle a (rare?) race condition in which during the intervening time between calling `SubscriptionState.fetchablePartitions()` and `SubscriptionState.position()` the partition was unassigned. I'm not sure why the existing code opted to throw an exception vs. "simply" logging some debug and `continue`-ing 🤔

The code that appears in the `FetchRequestManagerTest.testFetchRequestWithBufferedPartitionMissingPosition()` method is intentionally contrived in order to force that state to occur so we can verify it does throw the exception.

Reply:
Thanks for the explanation. Sounds good.